packages/server/src/shared/internal/serialize.ts

```diff
@@ -44,25 +43,22 @@ type SerializeTuple<T extends [unknown, ...unknown[]]> = {
 
 /** JSON serialize objects (not including arrays) and classes */
 type SerializeObject<T extends object> = {
-  [k in keyof Omit<T, FilterKeys<T, NonJsonPrimitive>>]: Serialize<T[k]>;
+  [k in keyof T as T[k] extends NonJsonPrimitive ? never : k]: Serialize<T[k]>;
```

Comment:
We need the indirection here to retain jump to definition. Using key remapping breaks jump to definition.

Reply:
Thanks @sachinraja. That is interesting. The jump to definition doesn't seem to work in WebStorm for both cases, but indeed in VSCode with the new code, it breaks. Wouldn't know how to add a test for that. But I'll see if there is a way to fix this without breaking jump to definition.

Reply:
Ok, fiddled around a bit in TS playground to see if we can have it working both ways, but I am afraid it does not seem to be possible. I am not sure why VSCode is able to jump to definition when using Omit instead (maybe someone can elaborate on this?). The limitation of Omit however - if I understand correctly - is that in the case of having a mixed object with known and unknown keys, the `keyof` collapses it into `string`, so the information of the known keys is lost. 

Not sure how to proceed with this but would see a number of options:
1. Prioritize jump to definition: in this case, do nothing.
2. Prioritize 'correct' types: apply the change.
3. Discourage users to implement types like this in the documentation. Personally, while working on this issue, I was pondering the use cases, especially when using tRPC, when would you have unknown and known keys mixed. The whole point of tRPC is to have typesafety.

Reply:
We need to keep jump to definition as it's a big part of "the tRPC experience." VSCode works when just using `Omit` because the TS server knows you aren't changing the keys themselves (but it thinks the keys themselves are changing if you use key remapping).

I would prioritize jump to definition as most users do not have unknown keys.